icu4c/source/i18n/measfmt.cpp

```diff
@@ -770,71 +715,96 @@ static void appendRange(
-    // we can replace it later so that we get '0:00:09.3' instead of
-    // '0:00:9.3'
-    FieldPosition intFieldPosition(UNUM_INTEGER_FIELD);
-    (*numberFormat)->format(
-            smallestAmount, smallestAmountFormatted, intFieldPosition, status);
-    if (
-            intFieldPosition.getBeginIndex() == 0 &&
-            intFieldPosition.getEndIndex() == 0) {
+
+    UnicodeString pattern;
+
+    double hours = hms[0].getDouble(status);
+    double minutes = hms[1].getDouble(status);
+    double seconds = hms[2].getDouble(status);
+
+    // All possible combinations: "h", "m", "s", "hm", "hs", "ms", "hms"
+    if (bitMap == 5 || bitMap == 7) { // "hms" & "hs" (we add minutes if "hs")
+        pattern = cache->getNumericDateFormatters()->hourMinuteSecond;
+        hours = uprv_trunc(hours);
+        minutes = uprv_trunc(minutes);
+    } else if (bitMap == 3) { // "hm"
+        pattern = cache->getNumericDateFormatters()->hourMinute;
+        hours = uprv_trunc(hours);
+    } else if (bitMap == 6) { // "ms"
+        pattern = cache->getNumericDateFormatters()->minuteSecond;
+        minutes = uprv_trunc(minutes);
+    } else { // h m s, handled outside formatNumeric. No value is also an error.
         status = U_INTERNAL_PROGRAM_ERROR;
         return appendTo;
     }
 
-    // Format time. draft becomes something like '5:30:45'
-    // #13606: DateFormat is not thread-safe, but MeasureFormat advertises itself as thread-safe.
-    FieldPosition smallestFieldPosition(smallestField);
-    UnicodeString draft;
-    static UMutex dateFmtMutex;
-    umtx_lock(&dateFmtMutex);
-    dateFmt.format(date, draft, smallestFieldPosition, status);
-    umtx_unlock(&dateFmtMutex);
-
-    // If we find field for smallest amount replace it with the formatted
-    // smallest amount from above taking care to replace the integer part
-    // with what is in original time. For example, If smallest amount
-    // is 9.35s and the formatted time is 0:00:09 then 9.35 becomes 09.35
-    // and replacing yields 0:00:09.35
-    if (smallestFieldPosition.getBeginIndex() != 0 ||
-            smallestFieldPosition.getEndIndex() != 0) {
-        appendRange(draft, 0, smallestFieldPosition.getBeginIndex(), appendTo);
-        appendRange(
-                smallestAmountFormatted,
-                0,
-                intFieldPosition.getBeginIndex(),
-                appendTo);
-        appendRange(
-                draft,
-                smallestFieldPosition.getBeginIndex(),
-                smallestFieldPosition.getEndIndex(),
-                appendTo);
-        appendRange(
-                smallestAmountFormatted,
-                intFieldPosition.getEndIndex(),
-                appendTo);
-        appendRange(
-                draft,
-                smallestFieldPosition.getEndIndex(),
-                appendTo);
-    } else {
-        appendTo.append(draft);
+    const NumberFormat *numberFormatter = numberFormat->get();
```

Comment:
Please look at `formatMeasure` for how to get the cached LocalizedNumberFormatter.  It is inefficient to use the public-facing DecimalFormat APIs in ICU implementation code.

Reply:
I'm not sure I understand.

I use the existing formatter, I don't create a new one.

**Me**
```
const SharedNumberFormat *numberFormat;
...
UnicodeString &MeasureFormat::formatNumeric( ... ) {
    ...
    const NumberFormat *numberFormatter = numberFormat->get();
    ...
}
```

**formatMeasure does this:**
```
UnicodeString &MeasureFormat::formatMeasure( ... const NumberFormat &nf ... ) {
    auto* df = dynamic_cast<const DecimalFormat*>(&nf);
    // check df == nullptr) { ... }
    number::FormattedNumber result;
    if (auto* lnf = df->toNumberFormatter(status)) {
        result = lnf->unit(amtUnit)
            .unitWidth(getUnitWidth(fWidth))
            .formatDouble(amtNumber.getDouble(status), status);
    }
}
```

So it takes a `NumberFormat` as parameter, then casts it to `DecimalFormat`, then gets a `toNumberFormatter` from it.

How does it get invoked (so that we know where the parameter comes from)?

A. in **MeasureFormat::format** 
```    formatMeasure(*amount, **numberFormat, appendTo, pos, status);```
B. in **MeasureFormat::formatMeasures**
```     formatMeasure(measures[0], **numberFormat, appendTo, pos, status);```
C. in **MeasureFormat::formatMeasures**
```     formatMeasure(measures[i], *nf, results[i], pos, status);
D. in **MeasureFormat::formatMeasuresSlowTrack**
```     formatMeasure(measures[i], *nf, results[i], fpos, status);
     formatMeasure(measures[i], *nf, results[i], dontCare, status);```

In A and B the `**numberFormat` is the same one I am using. No caching.
In C there is the cached integer formatter, sometimes
```
        const NumberFormat *nf = cache->getIntegerFormat();
        if (i == measureCount - 1) {
            nf = numberFormat->get();
        }
```
In D is the cached integer formatter: `const NumberFormat *nf = cache->getIntegerFormat();`

So C / D get use some caching, but that is the `getIntegerFormat`, which is the wrong formatter.
I don't want an integer formatter, because I needs to do stuff like 12:34.5678 (with decimals)

Reply:
The field `numberFormat` is the right one to use.  I'm just saying that it is more efficient to call `.toNumberFormatter()` and then use `.setIntegerWidth()` than it is to call `.clone()` and `.setMinimumIntegerDigits()`.

Reply:
Something like this:

```cpp
const auto* df = dynamic_cast<const DecimalFormat*>(numberFormat->get());
// handle nullptr
LocalizedNumberFormatter lnf2;
if (auto* lnf = df->toNumberFormatter(status)) {
    lnf2 = lnf->integerWidth(IntegerWidth::zeroFillTo(2));
} else {
    // return
}
```

It's more efficient because you don't generally have to allocate heap memory, and also note that `.setMinimumIntegerDigits()` under the hood does a bit of extra work that is avoided here.

Reply:
Done. Thanks a lot.